File: 09 September/09-23/stringMirror.js
function isMirror(str1, str2) {

  return str1.replace(/[^a-zA-Z]/g, "") === 
    str2.replace(/[^a-zA-Z]/g, "")
        .split("")
        .reverse()
        .join("");
}

console.log(isMirror("helloworld", "helloworld"));
console.log(isMirror("Hello World", "dlroW olleH"));
console.log(isMirror("RaceCar", "raCecaR"));
console.log(isMirror("RaceCar", "RaceCar"));
console.log(isMirror("Mirror", "rorrim"));
console.log(isMirror("Hello World", "dlroW-olleH"));
console.log(isMirror("Hello World", "!dlroW !olleH"));

// --- TEST SUITE ---

const testsText = `
1. isMirror("helloworld", "helloworld") should return false.
2. isMirror("Hello World", "dlroW olleH") should return true.
3. isMirror("RaceCar", "raCecaR") should return true.
4. isMirror("RaceCar", "RaceCar") should return false.
5. isMirror("Mirror", "rorrim") should return false.
6. isMirror("Hello World", "dlroW-olleH") should return true.
7. isMirror("Hello World", "!dlroW !olleH") should return true.
`;

const testsRegex = /(?<number>\d+)\.\s(?<functionCall>.+) should return (?<output>.+?)\.?$/gm;
const testData = [...testsText.matchAll(testsRegex).map(match => match.groups)];

function runTests(testData) {
  console.log("--------------------------");
  console.log("🧪Starting Verification...");
  console.log("--------------------------");

  let allPassed = true;
  let failCount = 0;
  
  testData.forEach(test => {
    const functionCallOutput = eval(test.functionCall);
    const testOutput = eval(test.output);
    let comparison;

    if (Array.isArray(testOutput))
      comparison = arraysEqual(functionCallOutput, testOutput);
    else
      comparison = functionCallOutput === testOutput;

    if (comparison) {
      console.log(`${test.number}.✅PASS - Function Call: ${test.functionCall}`);
    }
    else {
      console.log(`${test.number}.❌FAIL - Function Call: ${test.functionCall}\nExpected: ${testOutput}\nGot: ${functionCallOutput}`);
      allPassed = false;
      failCount ++;
    }
  })

  console.log("----------------------------");

  if (allPassed)
    console.log("🎉SUCCESS: All tests PASSED.");
  else
    console.log(`⚠️WARNING: ${failCount}/${testData.length} tests FAILED.`);
}

function arraysEqual(a, b) {
  if (a.length !== b.length)
    return false;

  for (const index in a) {
    if (Array.isArray(a[index])) {
      if (Array.isArray(b[index])) {
        if (!arraysEqual(a[index], b[index]))
          return false;
      }
      else 
        return false;
    }

    else if (a[index] !== b[index])
      return false;
  }

  return true;
}

runTests(testData);

File: 09 September/09-23/stringMirror.py
def is_mirror(str1, str2):
    import re

    return re.sub(r"[^a-zA-Z]", "", str1) == re.sub(r"[^a-zA-Z]", "", str2)[::-1]

print(is_mirror("helloworld", "helloworld"))
print(is_mirror("Hello World", "dlroW olleH"))
print(is_mirror("RaceCar", "raCecaR"))
print(is_mirror("RaceCar", "RaceCar"))
print(is_mirror("Mirror", "rorrim"))
print(is_mirror("Hello World", "dlroW-olleH"))
print(is_mirror("Hello World", "!dlroW !olleH"))

# --- TEST SUITE ---

tests_text = '''
1. is_mirror("helloworld", "helloworld") should return False.
2. is_mirror("Hello World", "dlroW olleH") should return True.
3. is_mirror("RaceCar", "raCecaR") should return True.
4. is_mirror("RaceCar", "RaceCar") should return False.
5. is_mirror("Mirror", "rorrim") should return False.
6. is_mirror("Hello World", "dlroW-olleH") should return True.
7. is_mirror("Hello World", "!dlroW !olleH") should return True.
'''

import re

tests_regex = re.compile(r"(?P<number>\d+)\.\s(?P<function_call>.+) should return (?P<output>.+?)\.?$", re.MULTILINE)
test_data = [item.groupdict() for item in re.finditer(tests_regex, tests_text)]

def run_tests(test_data):
    print("--------------------------")
    print("🧪 Starting Verification...")
    print("--------------------------")

    all_passed = True
    fail_count = 0

    for test in test_data:
        function_call_output = eval(test['function_call'])
        test_output = eval(test['output'])
        
        if function_call_output == test_output:
            print(f"{test['number']}.✅ PASS - Function Call:\n{test['function_call']}")
        else:
            print(f"{test['number']}.❌ FAIL - Function Call:\n{test['function_call']}\nExpected:\n{test_output}\nGot:\n{function_call_output}")
            all_passed = False
            fail_count += 1

    print("----------------------------");

    if (all_passed):
        print("🎉 SUCCESS: All tests PASSED.");
    else:
        print(f"⚠️ WARNING: {fail_count}/{len(test_data)} tests FAILED.");

run_tests(test_data)